server/lib/scheduler/Schedule.ts

```diff
@@ -0,0 +1,142 @@
+  }
+
+  private getUserSchedule(userId) {
+    return this.schedule.get(userId);
+  }
+
+  public async assignWorkspaceToUser(userId, workspace, startAtTimestamp = Date.now()) {
+    if (!this.doesUserHaveASchedule(userId)) {
+      this.createUserScheduleIfNotCreated(userId);
+    }
+
+    const userSchedule = this.getUserSchedule(userId);
+    userSchedule.assignWorkspace(workspace, startAtTimestamp);
+    const rootParent = await this.rootParentCache.getRootParentOfWorkspace(workspace);
+    this.cacheForWhenTreeLastWorkedOn[rootParent.id] = startAtTimestamp;
+  }
+
+  public getMostRecentAssignmentForUser(userId) {
+    if (!this.doesUserHaveASchedule(userId)) {
+      return undefined;
+    }
+
+    const userSchedule = this.getUserSchedule(userId);
+    return userSchedule.getMostRecentAssignment();
+  }
+
+  /*
+    NOTE: this only considers the trees associated with the rootWorkspaces
+    argument
+  */
+  public async isInTreeWorkedOnLeastRecently(rootWorkspaces, workspace) {
+    const treesWorksOnLeastRecently = this.getTreesWorkedOnLeastRecently(rootWorkspaces);
+    const rootParent = await this.rootParentCache.getRootParentOfWorkspace(workspace);
+    return !!(treesWorksOnLeastRecently.find(rootWorkspace => rootWorkspace === rootParent));
+  }
+
+  /*
+    NOTE: this only considers the trees associated with the rootWorkspaces
+    argument
+  */
+  public getTreesWorkedOnLeastRecently(rootWorkspaces) {
+    const treesNotYetWorkedOn = rootWorkspaces.filter(
+      rootWorkspace => this.cacheForWhenTreeLastWorkedOn[rootWorkspace.id] === undefined
+    );
+
+    if (treesNotYetWorkedOn.length > 0) {
+      return treesNotYetWorkedOn;
+    }
+
+    const lastWorkedOnTimestamps = rootWorkspaces.map(
+      rootWorkspace => this.cacheForWhenTreeLastWorkedOn[rootWorkspace.id]
+    );
+
+    const minTimestamp = Math.min.apply(Math, lastWorkedOnTimestamps);
+
+    const leastRecentlyWorkedOnTrees = rootWorkspaces.filter(rootWorkspace =>
+      this.cacheForWhenTreeLastWorkedOn[rootWorkspace.id] === minTimestamp
+    );
+
+    return leastRecentlyWorkedOnTrees;
+  }
+
+  public isWorkspaceCurrentlyBeingWorkedOn(workspace) {
+    for (const [userId, userSchedule] of this.schedule) {
+      if (userSchedule.isUserCurrentlyWorkingOnWorkspace(workspace)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  public hasWorkspaceBeenWorkedOnYet(workspace) {
+    for (const [userId, userSchedule] of this.schedule) {
+      if (userSchedule.hasUserWorkedOnWorkspace(workspace)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  public getTimestampWorkspaceLastWorkedOn(workspace) {
+    let mostRecentTimestamp = -Infinity;
+
+    for (const [userId, userSchedule] of this.schedule) {
+      const curTimestamp = userSchedule.getTimestampWorkspaceLastWorkedOn(workspace);
+      if (mostRecentTimestamp < curTimestamp) {
+        mostRecentTimestamp = curTimestamp;
+      }
+    }
```

Comment:
the above 8 lines are so should really be changed into a _.min function or similar.

Reply:
I'll do this now too.